jobs.py

```diff
@@ -228,185 +204,157 @@ def get_next_job(self, lifo=False, device=None, worker=None):
-                            'from jobs where device=? order by istry desc, '
-                            'created %s' % order,
-                            (device,))]
-            if not jobs:
-                return None
-            next_job = jobs[0]
-            next_job['attempts'] += 1
-            next_job['last_attempt'] = datetime.datetime.now().isoformat()
-            c.execute('update jobs set attempts=?, last_attempt=? where id=?',
-                      (next_job['attempts'], next_job['last_attempt'],
-                       next_job['id']))
-            conn.commit()
-            next_job['tests'] = []
-            test_rows = [
-                {'name': test_row[0], 'guid': test_row[1]}
-                 for test_row in
-                 c.execute('select name, guid from tests where jobid=?',
-                           (next_job['id'],))]
-            for test_row in test_rows:
-                # Convert the list of test names in the job to a list
-                # of test objects and update their guid values.
-                for test in worker.tests:
-                    if test.name == test_row['name']:
-                        test.job_guid = test_row['guid']
-                        next_job['tests'].append(test)
-        except sqlite3.OperationalError:
-            logger.exception('jobs.get_next_job')
-            next_job = None
-        if next_job:
-            logger.debug('jobs.get_next_job: %s' % next_job)
-        return next_job
-
-    def cancel_job(self, test_name, test_guid, device=None):
-        logger.debug('jobs.cancel_job: %s %s %s' % (test_name, test_guid, device))
+        conn = self._conn()
+
+        # Find the ids of the jobs whose attempts exceed the maximum.
+        # First delete the associated tests, then the jobs.
+        job_cursor = self._execute_sql(
+            conn,
+            'select id from jobs where device=? and attempts>=?',
+            values=(device, self.MAX_ATTEMPTS))
+        job_ids = [job[0] for job in job_cursor]
+        for job_id in job_ids:
+            self._execute_sql(conn, 'delete from tests where jobid=?',
+                              values=(job_id,))
+        job_cursor.close()
+        self._execute_sql(
+            conn,
+            'delete from jobs where device=? and attempts>=?',
+            values=(device, self.MAX_ATTEMPTS))
+
+        self._commit_connection(conn)
+
+        job_cursor = self._execute_sql(
+            conn,
+            'select id,created,last_attempt,build_url,'
+            'build_id,changeset,tree,revision,revision_hash,'
+            'enable_unittests,attempts,instr(build_url,"try") as istry '
+            'from jobs where device=? order by istry desc, '
+            'created %s' % order,
+            values=(device,))
+
+        job_row = job_cursor.fetchone()
+        job_cursor.close()
+        if not job_row:
+            self._close_connection(conn)
```

Comment:
The _conn, _execute_sql, _commit_connection, _close_connection,  methods won't raise any SQL OperationalError exceptions. They won't return until the statement completes error free.

It is hard to tell from the pysqlite documentation what the cursor methods might raise, but the cursor.close() and cursor.fetchone aren't wrapped in a try block so could potentially raise something. I think they are pretty safe though.

I'd rather not be limited to  

conn = self._conn()
// do everything you need to do
self._close_connection(conn)

for some of the longer functions since I think early returns keeps it more readable. It might be something we can do in the shorter ones though.

We could do 

try:
_use_conn(conn)
finally:
self._close_connection(conn)

if you don't mind increasing the nesting in the longer methods like get_next_job, or cancel_test.

I'll look at it some more and see what I can see.